docs/madr/decisions/031-automatic-rechable-services.md

```diff
@@ -0,0 +1,289 @@
+
+2. `MeshService/Mesh`  
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: MeshTrafficPermission
+metadata:
+  namespace: kuma-system
+  name: mtp-ms-to-mesh
+spec:
+  targetRef:
+    kind: MeshService
+    name: a
+  from:
+    - targetRef:
+        kind: Mesh
+      default:
+        action: Allow
+```
+
+```mermaid
+graph TD
+    B-->A
+    C-->A
+    D-->A
+```
+
+We could either define an edge between every service to that service or have a special list that contains allIncomingAllowed.
+Services from that list would always be added as an outgoing edge.
+
+- `Mesh/MeshService`
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: MeshTrafficPermission
+metadata:
+  namespace: kuma-system
+  name: mtp-mesh-to-ms
+spec:
+  targetRef:
+    kind: Mesh
+  from:
+    - targetRef:
+        kind: MeshService
+        name: a
+      default:
+        action: Allow
+```
+
+```mermaid
+graph TD
+    A-->B
+    A-->C
+    A-->D
+```
+
+This is simply `A` can talk to anyone, this could be either adding and outgoing edge from this service to every other service or having it on a list called allOutgoing and not bothering to trim the configuration.
+
+3. `MeshService/MeshService` - already talked about, this is the easy case
+
+4. Subsets 
+
+There is a strong case that makes sense using `k8s.kuma.io/namespace: xyz` to allow only traffic from a specific namespace.
+
+We will support `MeshSubset` and `MeshServiceSubset` on the `from` level and support `k8s.kuma.io/*` labels on the `top` level.
+
+#### Migration
+
+A typical migration to Mesh involves bringing services gradually into the Mesh.
+If the system is small enough, after several iterations, it will be enabled for the whole system (every service).
+
+If the system is too big to do this we need to do extra work in the iterations (defining MeshTrafficPermissions along the way).
+
+1. 0 services are in the Mesh
+2. Bring X services into the Mesh (where X is small enough, so the Mesh can handle it)
+3. Enable MeshTrafficPermission for services brought into the mesh (the performance improves because we trimmed the configs)
+4. Repeat point 2-3 until all services are in the Mesh
+
+The good thing about how this migration is set up is that we can gradually do it without any interruption to the system.
+
+It's good to keep in mind that the solution does not have to be perfect from the get-go.
+People can start with:
+
+```yaml
+apiVersion: kuma.io/v1alpha1
+kind: MeshTrafficPermission
+metadata:
+  namespace: kuma-system
+  name: namespace-controlled
+spec:
+  targetRef:
+    kind: MeshSubset
+    tags:
+      namespace: my-ns
+  from:
+    - targetRef:
+        kind: MeshService
+        name: a
+      default:
+        action: Allow
+
+```
+
+The downside of it is that  to avoid any disruptions.
+
+The migration assumes using Kubernetes DNS.
+
+Let's take the following service graph as an example:
+
+```mermaid
+flowchart LR
```

Comment:
This graph doesn't show the situation when the service has at least 2 different clients. I think it will be useful to cover migration for such case as well. Also, worth mentioning mTLS mode, service migration should work both for Disabled and Permissive mTLS.

Reply:
Will add this and a section on mTLS